graphql_compiler/cost_estimation/cardinality_estimator.py

```diff
@@ -82,57 +82,126 @@ def _get_parent_and_child_name_from_edge(schema_graph, child_location):
+    _, edge_name = _get_last_edge_direction_and_name_to_location(child_location)
+    outbound_vertex_name, inbound_vertex_name = _get_outbound_and_inbound_names_from_locations(
+        query_metadata, child_location, parent_location
+    )
+    probe_result = statistics.get_edge_count_between_vertex_pair(
+        outbound_vertex_name, inbound_vertex_name, edge_name
+    )
+    return probe_result
 
-    Given a parent location of type A and child location of type B, assume all AB edges are
-    distributed evenly over A vertices, so the expected number of child edges per parent vertex is
-    (# of AB edges) / (# of A vertices). If A and B are subclasses of C and D respectively, we only
-    have access to CD edges, so in general, we'll use (# of CD edges) / (# of C vertices), but since
-    we're only interested in edges to Bs, we'll scale this result by the fraction of Bs over Ds.
+
+def _estimate_edge_count_between_vertex_pair_using_class_count(
+    schema_graph, statistics, query_metadata, child_location, parent_location
+):
+    """Estimate the count of edges connecting two vertex classes via the class_count statistic.
+
+    Given a parent location of type A and a child location of type B, we estimate the number of AB
+    edges using class counts. If A and B are subclasses of C and D respectively, we only have
+    access to CD edges, so in general, we assume that the number of CD edges with A as an endpoint
+    are proportional to the fraction of A vertices over C vertices, and likewise for B and D.
+    Using this assumption, in general we estimate the statistic as
+    (# of AB edges) = (# of CD edges) * (# of A vertices) / (# of C vertices) *
+                                        (# of B vertices) / (# of D vertices).
 
     Args:
         schema_graph: SchemaGraph object
         statistics: Statistics object
         query_metadata: QueryMetadataTable object
-        parameters: dict, parameters with which query will be executed
         child_location: BaseLocation object
+        parent_location: BaseLocation object
 
     Returns:
-        float, expected number of child_location vertices connected to each parent_location vertex.
+        float, estimate for (# of AB edges)
     """
-    # Count the number of edges between child_location and parent_location type vertices.
     _, edge_name = _get_last_edge_direction_and_name_to_location(child_location)
-    # TODO(evan): If edge is recursed over, we need a more detailed statistic
-    # TODO(vlad): Use get_edge_count_between_vertex_pair statistic
     edge_counts = statistics.get_class_count(edge_name)
 
-    # Scale edge_counts if child_location's type is a subclass of the edge's endpoint type.
+    parent_name_from_location = query_metadata.get_location_info(parent_location).type.name
+    child_name_from_location = query_metadata.get_location_info(child_location).type.name
     parent_name_from_edge, child_name_from_edge = _get_parent_and_child_name_from_edge(
```

Comment:
Let's rename these variables to `child_` and `parent_base_classname` to make it clear that these names might be more general than the names we get from the locations themselves in `_get_outbound_and_inbound_names_from_locations` (e.g. `Entity` instead of `Bank`). Same for the `_get_parent_and_child_name_from_edge` method above.